File: PrintPulse.Core/BambuApi.cs
using System.Net;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;

namespace PrintPulse.Core;

public sealed class CloudException(string message, bool expired = false) : Exception(message) { public bool Expired { get; } = expired; }
public sealed record LoginResult(Session? Session, string? Challenge, string? TfaKey);
public sealed class BambuApi : IDisposable
{
    private readonly HttpClient http;
    private readonly CookieContainer cookies = new();
    public BambuApi(HttpMessageHandler? handler = null)
    {
        http = new HttpClient(handler ?? new HttpClientHandler { CookieContainer = cookies, AutomaticDecompression = DecompressionMethods.All }) { Timeout = TimeSpan.FromSeconds(25) };
        http.DefaultRequestHeaders.UserAgent.ParseAdd("PrintPulse/1.0");
        http.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
    }
    public static string Root(bool china) => china ? "https://api.bambulab.cn" : "https://api.bambulab.com";
    private async Task<JsonElement> Request(string url, object? body, string? token, CancellationToken ct)
    {
        using var request = new HttpRequestMessage(body == null ? HttpMethod.Get : HttpMethod.Post, url);
        if (body != null) request.Content = JsonContent.Create(body);
        if (token != null) request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        using var response = await http.SendAsync(request, ct).ConfigureAwait(false);
        if (response.StatusCode == HttpStatusCode.Unauthorized) throw new CloudException("Session expired. Sign in again in Settings.", true);
        if (response.StatusCode == HttpStatusCode.Forbidden) throw new CloudException("Bambu blocked this sign-in method. Try email-code sign-in; a browser verification may be required.");
        if ((int)response.StatusCode == 429) throw new CloudException("Bambu is limiting requests. Wait a few minutes before trying again.");
        if (!response.IsSuccessStatusCode) throw new CloudException(body == null ? "Bambu Cloud is unavailable. PrintPulse will retry." : "Sign-in was rejected. Check the details or request a new code.");
        try { using var doc = JsonDocument.Parse(await response.Content.ReadAsStringAsync(ct)); return doc.RootElement.Clone(); }
        catch (JsonException) { throw new CloudException("Bambu returned an unexpected response. Try again later."); }
    }
    public async Task<LoginResult> Login(string email, string value, bool code, bool china, CancellationToken ct)
    {
        object body = code ? new { account = email, code = value } : new { account = email, password = value, apiError = "" };
        var result = await Request(Root(china) + "/v1/user-service/user/login", body, null, ct);
        if (Printer.Text(result, "accessToken") is { Length: > 0 } token)
            return new(await MakeSession(token, email, china, ct), null, null);
        return Printer.Text(result, "loginType") switch
        {
            "verifyCode" => new(null, "email", null),
            "tfa" => new(null, "tfa", Printer.Text(result, "tfaKey")),
            _ => throw new CloudException("Sign-in was not completed. Check your code or try email-code sign-in.")
        };
    }
    public async Task SendCode(string email, bool china, CancellationToken ct) => _ = await Request(Root(china) + "/v1/user-service/user/sendemail/code", new { email, type = "codeLogin" }, null, ct);
    public async Task<Session> VerifyTfa(string key, string code, string email, bool china, CancellationToken ct)
    {
        var root = china ? "https://bambulab.cn" : "https://bambulab.com";
        using var request = new HttpRequestMessage(HttpMethod.Post, root + "/api/sign-in/tfa") { Content = JsonContent.Create(new { tfaKey = key, tfaCode = code }) };
        using var response = await http.SendAsync(request, ct).ConfigureAwait(false);
        if (response.StatusCode == HttpStatusCode.Forbidden) throw new CloudException("Bambu requires browser verification for authenticator sign-in. Try email-code sign-in.");
        if (!response.IsSuccessStatusCode) throw new CloudException("Authenticator code was rejected. Try the current code or email-code sign-in.");
        // The TFA endpoint establishes a cookie session; its body need not be JSON.
        var token = cookies.GetCookies(new Uri(root))["token"]?.Value;
        if (string.IsNullOrWhiteSpace(token)) throw new CloudException("Authenticator sign-in did not return a session. Try email-code sign-in.");
        return await MakeSession(token, email, china, ct);
    }
    private async Task<Session> MakeSession(string token, string email, bool china, CancellationToken ct)
    {
        string? username = null;
        try
        {
            var part = token.Split('.')[1].Replace('-', '+').Replace('_', '/');
            using var doc = JsonDocument.Parse(Convert.FromBase64String(part.PadRight((part.Length + 3) / 4 * 4, '=')));
            username = Printer.Text(doc.RootElement, "username");
        } catch (Exception e) when (e is FormatException or JsonException or IndexOutOfRangeException) { }
        if (string.IsNullOrWhiteSpace(username))
        {
            var result = await Request(Root(china) + "/v1/design-user-service/my/preference", null, token, ct);
            username = Printer.Text(result, "uid") is {} uid ? "u_" + uid : null;
        }
        if (string.IsNullOrWhiteSpace(username)) throw new CloudException("Bambu did not provide an account identifier. Please sign in again.");
        return new Session(token, username, email, china);
    }
    public async Task<JsonElement[]> Discover(Session session, CancellationToken ct)
    {
        var result = await Request(Root(session.China) + "/v1/iot-service/api/user/bind", null, session.Token, ct);
        if (result.ValueKind != JsonValueKind.Object || !result.TryGetProperty("devices", out var devices) || devices.ValueKind != JsonValueKind.Array) throw new CloudException("Bambu did not return a printer list. PrintPulse will retry.");
        return devices.EnumerateArray().Select(d => d.Clone()).ToArray();
    }
    public async Task<JsonElement[]> Tasks(Session session, CancellationToken ct)
    {
        var result = await Request(Root(session.China) + "/v1/user-service/my/tasks?limit=100", null, session.Token, ct);
        if (result.ValueKind != JsonValueKind.Object) throw new CloudException("Bambu task metadata is temporarily unavailable.");
        return result.TryGetProperty("hits", out var hits) && hits.ValueKind == JsonValueKind.Array ? hits.EnumerateArray().Select(d => d.Clone()).ToArray() : [];
    }
    public void Dispose() => http.Dispose();
}

File: ios/Shared/BambuAPI.swift
import Foundation

enum CloudError: LocalizedError {
    case expired, rejected, limited, unavailable, malformed, verification
    var errorDescription: String? {
        switch self {
        case .expired: return "Session expired. Open PrintPulse and sign in again."
        case .rejected: return "Bambu rejected this request. Check your sign-in details or try email-code sign-in."
        case .limited: return "Bambu is limiting requests. Wait a few minutes before retrying."
        case .unavailable: return "Bambu Cloud is unavailable. Showing the last update."
        case .malformed: return "Bambu returned an unexpected response. Please try again later."
        case .verification: return "Bambu requires browser verification. Try email-code sign-in."
        }
    }
}

enum LoginResult {
    case session(CloudSession), emailCode, authenticator(String)
}

final class BambuAPI {
    private let network: URLSession
    init() {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 10
        config.timeoutIntervalForResource = 15
        config.httpAdditionalHeaders = ["User-Agent": "PrintPulse-iOS/1.0", "Accept": "application/json"]
        network = URLSession(configuration: config)
    }
    deinit { network.invalidateAndCancel() }
    static func root(_ china: Bool) -> String { china ? "https://api.bambulab.cn" : "https://api.bambulab.com" }
    private func request(_ url: String, body: [String: String]? = nil, token: String? = nil, timeout: TimeInterval = 10) async throws -> [String: Any] {
        var request = URLRequest(url: URL(string: url)!)
        request.timeoutInterval = timeout
        if let body {
            request.httpMethod = "POST"
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        let (data, response) = try await network.data(for: request)
        try Self.validate(response)
        guard data.count <= 5_000_000, let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw CloudError.malformed }
        return json
    }
    static func validate(_ response: URLResponse) throws {
        guard let response = response as? HTTPURLResponse else { throw CloudError.unavailable }
        switch response.statusCode {
        case 200..<300: break
        case 401: throw CloudError.expired
        case 403: throw CloudError.verification
        case 429: throw CloudError.limited
        case 400..<500: throw CloudError.rejected
        default: throw CloudError.unavailable
        }
    }
    func login(email: String, secret: String, code: Bool, china: Bool) async throws -> LoginResult {
        let body = code ? ["account": email, "code": secret] : ["account": email, "password": secret, "apiError": ""]
        let result = try await request(Self.root(china) + "/v1/user-service/user/login", body: body)
        if let token = text(result, "accessToken"), !token.isEmpty { return .session(try await makeSession(token: token, email: email, china: china)) }
        if text(result, "loginType") == "verifyCode" { return .emailCode }
        if text(result, "loginType") == "tfa", let key = text(result, "tfaKey") { return .authenticator(key) }
        throw CloudError.rejected
    }
    func sendCode(email: String, china: Bool) async throws {
        _ = try await request(Self.root(china) + "/v1/user-service/user/sendemail/code", body: ["email": email, "type": "codeLogin"])
    }
    func verifyAuthenticator(key: String, code: String, email: String, china: Bool) async throws -> CloudSession {
        let root = china ? "https://bambulab.cn" : "https://bambulab.com"
        var req = URLRequest(url: URL(string: root + "/api/sign-in/tfa")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: ["tfaKey": key, "tfaCode": code])
        let (_, response) = try await network.data(for: req)
        try Self.validate(response)
        guard let token = network.configuration.httpCookieStorage?.cookies(for: URL(string: root)!)?.first(where: { $0.name == "token" })?.value else { throw CloudError.verification }
        return try await makeSession(token: token, email: email, china: china)
    }
    private func makeSession(token: String, email: String, china: Bool) async throws -> CloudSession {
        let preference = try await request(Self.root(china) + "/v1/design-user-service/my/preference", token: token)
        guard let uid = text(preference, "uid"), !uid.isEmpty else { throw CloudError.malformed }
        return CloudSession(token: token, username: "u_" + uid, email: email, china: china)
    }
    func devices(_ session: CloudSession) async throws -> [[String: Any]] {
        let result = try await request(Self.root(session.china) + "/v1/iot-service/api/user/bind", token: session.token, timeout: 5)
        guard let devices = result["devices"] as? [[String: Any]] else { throw CloudError.malformed }
        return devices
    }
    func tasks(_ session: CloudSession) async throws -> [[String: Any]] {
        let result = try await request(Self.root(session.china) + "/v1/user-service/my/tasks?limit=100", token: session.token, timeout: 4)
        return result["hits"] as? [[String: Any]] ?? []
    }
}
